src/schedule/zephyr_preempt_domain.c

```diff
@@ -0,0 +1,460 @@
+{
+	int ret;
+
+	dma_interrupt_legacy(data->channel, DMA_IRQ_CLEAR);
+
+	ret = interrupt_register(data->irq, dma_irq_handler, data);
+	if (ret < 0)
+		return ret;
+
+	interrupt_enable(data->irq, data);
+
+	interrupt_clear_mask(data->irq, BIT(data->channel->index));
+
+	dma_interrupt_legacy(data->channel, DMA_IRQ_UNMASK);
+
+	return 0;
+}
+
+static int register_dma_irq(struct zephyr_preempt_domain *domain,
+			    struct zephyr_preempt_domain_data **data,
+			    struct zephyr_preempt_domain_thread *dt,
+			    struct pipeline_task *pipe_task,
+			    int core)
+{
+	struct dma *crt_dma;
+	struct dma_chan_data *crt_chan;
+	struct zephyr_preempt_domain_data *crt_data;
+	int i, j, irq, ret;
+
+	/* register the DMA IRQ only for PPL tasks marked as "registrable"
+	 *
+	 * this is done because, in case of mixer topologies there's
+	 * multiple PPLs having the same scheduling component so there's
+	 * no need to go through this function for all of those PPL
+	 * tasks - only the PPL task containing the scheduling component
+	 * will do the registering
+	 *
+	 */
+	if (!pipe_task->registrable)
+		return 0;
+
+	/* iterate through all available channels in order to find a
+	 * suitable channel for which the DMA IRQs will be enabled.
+	 */
+	for (i = 0; i < domain->num_dma; i++) {
+		crt_dma = domain->dma_array + i;
+
+		for (j = 0; j < crt_dma->plat_data.channels; j++) {
+			crt_chan = crt_dma->chan + j;
+			crt_data = &domain->data[i][j];
+
+			/* skip if channel is not a scheduling source */
+			if (!dma_is_scheduling_source(crt_chan))
+				continue;
+
+			/* skip if channel is not active */
+			if (crt_chan->status != COMP_STATE_ACTIVE)
+				continue;
+
+			/* skip if channel not owned by current core */
+			if (core != crt_chan->core)
+				continue;
+
+			/* skip if IRQ was already registered */
+			if (crt_data->enabled_irq)
+				continue;
+
+			/* get IRQ number for current channel */
+			irq = interrupt_get_irq(dma_chan_irq(crt_dma, j),
+						dma_chan_irq_name(crt_dma, j));
+
+			crt_data->irq = irq;
+			crt_data->channel = crt_chan;
+			crt_data->dt = dt;
+
+			if (dt->started) {
+				/* if the Zephyr thread was started, we
+				 * can safely enable the DMA IRQs
+				 */
+				ret = enable_dma_irq(crt_data);
+				if (ret < 0)
+					return ret;
+
+				crt_data->enabled_irq = true;
+			}
+
+			/* let caller know we have found a channel */
+			*data = crt_data;
+
+			return 0;
+		}
+	}
+
+	/* if this point is reached then that means we weren't able to
+	 * find a suitable channel, let caller know
+	 */
+	return -EINVAL;
+}
+
+static int zephyr_preempt_domain_register(struct ll_schedule_domain *domain,
+					  struct task *task,
+					  void (*handler)(void *arg),
+					  void *arg)
+{
+	struct zephyr_preempt_domain *zephyr_preempt_domain;
+	struct zephyr_preempt_domain_thread *dt;
+	struct pipeline_task *pipe_task;
+	struct zephyr_preempt_domain_data *data;
+	k_tid_t thread;
+	int core, ret;
+	char thread_name[] = "ll_thread0";
```

Comment:
Should this be const, i.e. should we base this on the pipeline name ?

Reply:
I'm not sure that would work for now as we can have multiple pipeline tasks running on the same Zephyr thread so I think a generic name is more suitable.